File: frontend/src/types/ifc.ts
export type IFCDataSource = "IFC_SOURCE" | "CALCULATED" | "AI_INFERRED" | "USER_ENTERED" | "MISSING";

export interface IFCModelGroup {
  id: string; projectId: string; name: string; discipline?: string; description?: string;
  activeVersionId?: string; baselineVersionId?: string; archivedAt?: string;
}

export interface IFCMetric { count: number; total: number; percentage: number }
export interface IFCGeoreferencing {
  status?: string; impact?: string; coordinateReferenceSystem?: string; epsgCode?: string;
  latitude?: number; longitude?: number; easting?: number; northing?: number; elevation?: number;
  orthogonalHeight?: number; mapConversion?: Record<string, unknown>; source?: IFCDataSource;
}
/**
 * Why the interference rule declined to run.
 *
 * Exactly the values `app/services/ifc_interference.py` and
 * `run_interference_analysis` can store. Each one means the model was *not*
 * checked, which is a different answer from "checked and clean" — the reason
 * the report carries them at all.
 */
export type IFCAnalysisSkipReason =
  | "UNKNOWN_LENGTH_UNIT"
  | "NO_ELEMENT_GEOMETRY_AVAILABLE"
  | "NO_STRUCTURAL_AND_SERVICE_PAIR_AVAILABLE"
  | "DISABLED_OR_MISSING_VERSION";

/** What the interference rule reports about its own run. */
export interface IFCInterferenceReport {
  method?: string;
  /** False when the rule declined to run; `skippedReason` then says why. */
  analysed?: boolean;
  skippedReason?: IFCAnalysisSkipReason | string | null;
  structuralElements?: number;
  serviceElements?: number;
  elementsWithoutGeometry?: number;
  comparisons?: number;
  lengthUnit?: string | null;
  truncated?: boolean;
  findingCount?: number;
  stored?: number;
  reviewedPairsPreserved?: number;
}

export interface IFCSummary {
  schema?: string; sites?: number; buildings?: number; storeys?: number; spaces?: number; zones?: number; elements?: number;
  projectOverview?: Record<string, unknown> & { sources?: Record<string, IFCDataSource> };
  mainStatistics?: Record<string, number>; modelCompleteness?: Record<string, IFCMetric>;
  disciplineBreakdown?: Record<string, number>; majorElementCategories?: Record<string, number>;
  hierarchyCounts?: Record<string, number>; georeferencing?: IFCGeoreferencing; units?: Record<string, string>;
  warnings?: string[]; revision?: string; versionType?: string; modelGroup?: string;
  processingStages?: Array<{ key: string; label: string; percentage: number }>;
  assetType?: { value?: string; confidence?: number; evidence?: string[] };
  spaceCategories?: Record<string, number>;
  /** Absent when interference analysis was never recorded for this revision. */
  interference?: IFCInterferenceReport;
  intelligenceSummary?: { text?: string; strengths?: string[]; missingInformation?: string[]; coordinationRisks?: string[]; recommendedNextSteps?: string[]; source?: string; reviewNotice?: string };
}

/** The server's upload rules, fetched so the client can check a file before sending it. */
export interface IFCUploadConstraints {
  maxFileBytes: number; maxFileMb: number; acceptedExtensions: string[];
  maxEntityCount: number; supportedSchemas: string[]; geometryEnabled: boolean;
}

export interface IFCVersion {
  id: string; modelGroupId: string; projectId: string; versionNumber: number;
  revisionCode?: string; versionType: string; title: string; description?: string; discipline?: string;
  originalFilename: string; fileHash: string; fileSize: number; ifcSchema?: string;
  authoringApplication?: string; processingStatus: string; processingProgress: number;
  processingDurationMs?: number; parsingErrorCode?: string; parsingErrorMessage?: string; supportLogId?: string;
  entityCount: number; geometryStatus: string;
  geometryError?: string; geometryStatsJson?: Record<string, unknown>; geometryGeneratedAt?: string;
  analysisStatus: string; modelSummaryJson: IFCSummary; isActive: boolean; createdAt: string;
  isBaseline: boolean; assetTypeSuggestion?: string; assetTypeConfidence?: number;
}

export interface IFCSpatialNode {
  id: string; versionId: string; globalId: string; entityType: string; name: string; description?: string;
  parentId?: string; nodeType: string; elevation?: number; area?: number; volume?: number; metadataJson?: Record<string, unknown>;
}

export interface IFCSpatialDetails {
  node: IFCSpatialNode; childCounts: Record<string, number>; elementCount: number;
  elementCategories: Record<string, number>; disciplines: Record<string, number>; relatedElementIds: string[];
  projectActivity: Record<string, number>; measurements: Record<string, { value: number; unit?: string; sourceType: string; sourceProperty?: string; confidence: number } | null>;
  spaceClassification?: { category: string; label: string; confidence: number; source: string; evidence?: string; method: string };
  availabilityNotice: string;
}

export interface IFCClassification { system?: string; code?: string; name?: string; source?: IFCDataSource }
export interface IFCElementMetadata {
  stepId?: number; originalName?: string; sourceGlobalId?: string; classifications?: IFCClassification[];
  disciplineSource?: IFCDataSource; disciplineConfidence?: number; disciplineReason?: string;
  disciplineSourceAttributes?: string[]; missingData?: string[]; completenessStatus?: "COMPLETE" | "PARTIAL" | "INCOMPLETE";
  materialStatus?: "DEFINED" | "VIRTUAL_REFERENCE" | "MISSING_FROM_SOURCE"; units?: Record<string, string>;
}
export interface IFCElement {
  id: string; versionId: string; globalId: string; entityType: string; name: string;
  description?: string; objectType?: string; predefinedType?: string; tag?: string; storeyNodeId?: string;
  spaceNodeId?: string; zoneNodeId?: string; buildingNodeId?: string; discipline?: string; systemName?: string; typeName?: string;
  materialSummary?: string; propertiesJson: Record<string, unknown>; quantitiesJson: Record<string, unknown>;
  boundingBoxJson?: Record<string, unknown>; geometryReference?: string; geometryHash?: string; placementHash?: string;
  metadataJson: IFCElementMetadata;
}

export interface IFCComparison {
  id: string; baseVersionId: string; targetVersionId: string; status: string;
  summaryJson: { total?: number; counts?: Record<string, number>; disciplineBreakdown?: Record<string, number>; comparisonConfidence?: string; unstableIdentifierCount?: number; confidenceMessage?: string };
  createdAt: string; error?: string;
}

/** One element as an interference finding describes it. */
export interface IFCInterferenceParty {
  elementId: string; globalId: string; name: string; entityType: string;
  discipline?: string; storey?: string; system?: string;
  boundingBox?: { min: number[]; max: number[] };
}

export interface IFCFinding {
  id: string; versionId: string; findingType: string; severity: string; title: string; description: string;
  discipline: string; disciplines: string[]; status: string; ifcRule: string; whyItMatters: string;
  recommendedAction: string; affectedElementIds: string[]; affectedElementCount: number; createdAt: string;
  /** How far the detection method can be trusted. Metadata-quality rules are 1. */
  confidence?: number;
  storey?: string; space?: string;
  /** Present on geometry-derived findings: what is being claimed, and how to check it. */
  claim?: string; verification?: string;
  /** Present on interference findings: the two elements and the measured overlap. */
  elementPair?: {
    structural?: IFCInterferenceParty; service?: IFCInterferenceParty;
    penetrationMetres?: number; overlapBox?: { min: number[]; max: number[] };
  };
}

/**
 * One page of a paged IFC collection.
 *
 * Mirrors the `IFCPaged` schema the backend already returns from
 * `/versions/{id}/elements` and `/versions/{id}/search`, which is this
 * project's pagination shape.
 */
export interface IFCPage<T> { items: T[]; total: number; page: number; pageSize: number }

/**
 * The review decisions `PATCH /projects/{id}/ifc/findings/{id}` accepts.
 *
 * Exactly the set `review_finding` in `app/api/ifc.py` allows; anything else
 * is rejected with 400. `ISSUE_CREATED` is deliberately absent — the server
 * sets that itself when a finding becomes an issue, and it is not a decision
 * a client may post.
 */
export type IFCFindingReviewStatus = "ACKNOWLEDGED" | "IGNORED" | "FALSE_POSITIVE";

/**
 * What a review call gives back.
 *
 * The three review routes return the ORM row directly — they declare no
 * `response_model` — so the payload is snake_case and not the camelCase shape
 * `GET /findings` assembles. Only the field the UI reads is declared, and it
 * is optional: the caller already knows which status it asked for and treats
 * this as confirmation rather than as the source of truth.
 */
export interface IFCFindingReviewResult { status?: string }

/** The issue `POST /findings/{id}/create-issue` opens, as that route returns it. */
export interface IFCFindingIssue { id: string; title: string; status: string }

export interface IFCSuggestion {
  id: string; versionId: string; suggestionType: string; payloadJson: Record<string, unknown>;
  title: string; discipline: string; priority: string; reason: string; affectedElementCount: number;
  expectedBenefit: string; recommendedAction: string; sourceFinding: string;
  confidence: number; status: string; aiInferred: boolean; createdAt: string;
}

File: backend/app/schemas/ifc.py
"""API contracts for the project-scoped IFC intelligence workspace."""
from datetime import date, datetime
from typing import Any
from uuid import UUID

from pydantic import ConfigDict, Field, model_validator

from app.schemas.user import CamelModel


class IFCCamelModel(CamelModel):
    model_config = ConfigDict(protected_namespaces=())


class IFCModelCreate(IFCCamelModel):
    name: str = Field(min_length=2, max_length=200)
    discipline: str | None = Field(default=None, max_length=50)
    description: str | None = Field(default=None, max_length=4000)


class IFCModelUpdate(IFCCamelModel):
    name: str | None = Field(default=None, min_length=2, max_length=200)
    discipline: str | None = Field(default=None, max_length=50)
    description: str | None = Field(default=None, max_length=4000)


class IFCModelOut(IFCCamelModel):
    id: UUID
    project_id: UUID
    name: str
    discipline: str | None = None
    description: str | None = None
    active_version_id: UUID | None = None
    baseline_version_id: UUID | None = None
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class IFCVersionOut(IFCCamelModel):
    id: UUID
    model_group_id: UUID
    project_id: UUID
    version_number: int
    revision_code: str | None = None
    version_type: str
    title: str
    description: str | None = None
    discipline: str | None = None
    authoring_source: str | None = None
    original_filename: str
    file_hash: str
    file_size: int
    ifc_schema: str | None = None
    authoring_application: str | None = None
    uploaded_by_id: UUID
    issue_date: date | None = None
    is_active: bool
    is_baseline: bool
    parent_version_id: UUID | None = None
    processing_status: str
    processing_progress: int
    parsing_error_code: str | None = None
    parsing_error_message: str | None = None
    support_log_id: str | None = None
    entity_count: int
    geometry_status: str
    geometry_error: str | None = None
    geometry_stats_json: dict[str, Any] = Field(default_factory=dict)
    geometry_generated_at: datetime | None = None
    analysis_status: str
    processing_duration_ms: int | None = None
    model_summary_json: dict[str, Any]
    asset_type_suggestion: str | None = None
    asset_type_confidence: float | None = None
    asset_type_confirmed: str | None = None
    row_version: int
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class IFCVersionPatch(IFCCamelModel):
    title: str | None = Field(default=None, min_length=1, max_length=250)
    description: str | None = Field(default=None, max_length=4000)
    revision_code: str | None = Field(default=None, max_length=80)
    version_type: str | None = Field(default=None, max_length=50)
    issue_date: date | None = None
    asset_type_confirmed: str | None = Field(default=None, max_length=80)
    row_version: int = Field(ge=1)


class IFCSpatialNodeOut(IFCCamelModel):
    id: UUID
    version_id: UUID
    global_id: str
    entity_type: str
    name: str
    description: str | None = None
    parent_id: UUID | None = None
    node_type: str
    elevation: float | None = None
    area: float | None = None
    volume: float | None = None
    metadata_json: dict[str, Any]


class IFCElementOut(IFCCamelModel):
    id: UUID
    version_id: UUID
    global_id: str
    entity_type: str
    name: str
    description: str | None = None
    object_type: str | None = None
    predefined_type: str | None = None
    tag: str | None = None
    storey_node_id: UUID | None = None
    space_node_id: UUID | None = None
    zone_node_id: UUID | None = None
    building_node_id: UUID | None = None
    discipline: str | None = None
    system_name: str | None = None
    type_name: str | None = None
    material_summary: str | None = None
    properties_json: dict[str, Any]
    quantities_json: dict[str, Any]
    bounding_box_json: dict[str, Any] | None = None
    geometry_reference: str | None = None
    geometry_hash: str | None = None
    placement_hash: str | None = None
    metadata_json: dict[str, Any]


class IFCComparisonCreate(IFCCamelModel):
    base_version_id: UUID
    target_version_id: UUID

    @model_validator(mode="after")
    def distinct_versions(self):
        if self.base_version_id == self.target_version_id:
            raise ValueError("Comparison versions must be different")
        return self


class IFCComparisonOut(IFCCamelModel):
    id: UUID
    project_id: UUID
    base_version_id: UUID
    target_version_id: UUID
    status: str
    summary_json: dict[str, Any]
    created_by_id: UUID
    error: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime


class IFCReviewRequest(IFCCamelModel):
    status: str = ""
    note: str | None = Field(default=None, max_length=4000)
    edited_payload: dict[str, Any] | None = None


class IFCBulkReview(IFCCamelModel):
    suggestion_ids: list[UUID] = Field(min_length=1, max_length=100)
    status: str
    note: str | None = Field(default=None, max_length=4000)


class IFCLinkCreate(IFCCamelModel):
    version_id: UUID
    ifc_element_id: UUID | None = None
    spatial_node_id: UUID | None = None
    linked_entity_type: str
    linked_entity_id: UUID
    link_type: str = Field(default="RELATED", max_length=60)

    @model_validator(mode="after")
    def one_ifc_source(self):
        if bool(self.ifc_element_id) == bool(self.spatial_node_id):
            raise ValueError("Provide exactly one IFC element or spatial node")
        return self


class IFCLinkOut(IFCCamelModel):
    id: UUID
    project_id: UUID
    version_id: UUID
    ifc_element_id: UUID | None = None
    spatial_node_id: UUID | None = None
    linked_entity_type: str
    linked_entity_id: UUID
    link_type: str
    confidence: float
    source: str
    confirmed_by_id: UUID | None = None
    confirmed_at: datetime | None = None
    created_at: datetime


class IFCPaged(IFCCamelModel):
    items: list[Any]
    total: int
    page: int
    page_size: int
